File: src/cleanup.test.ts
import path from "node:path";
import { test } from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs/promises";
import os from "node:os";
import { assessCleanupSafety, buildConfirmToken, evaluateCleanupGuards, executeCleanup } from "./cleanup.js";
import type { RepoScan } from "./scanner.js";

const row = (overrides: Partial<RepoScan> = {}): RepoScan => {
  const repoPath = "/tmp/demo-repo";
  return {
    repoPath,
    nodeModulesPath: path.join(repoPath, "node_modules"),
    manager: "yarn",
    hasLockfile: true,
    inactiveDays: 120,
    bytes: 1234,
    reinstallCommand: "yarn install --immutable",
    git: { branch: "main", dirty: false, ahead: 0, behind: 0 },
    ...overrides
  };
};

test("guard passes for repo/node_modules with lockfile", () => {
  const guard = evaluateCleanupGuards(row());
  assert.equal(guard.ok, true);
  assert.deepEqual(guard.reasons, []);
});

test("guard fails when lockfile missing", () => {
  const guard = evaluateCleanupGuards(row({ hasLockfile: false }));
  assert.equal(guard.ok, false);
  assert.match(guard.reasons.join(","), /missing lockfile/);
});

test("guard fails when target not node_modules", () => {
  const guard = evaluateCleanupGuards(row({ nodeModulesPath: "/tmp/demo-repo/dist" }));
  assert.equal(guard.ok, false);
  assert.match(guard.reasons.join(","), /target must be node_modules/);
});

test("guard fails when target is outside repo", () => {
  const guard = evaluateCleanupGuards(row({ nodeModulesPath: "/tmp/other/node_modules" }));
  assert.equal(guard.ok, false);
  assert.match(guard.reasons.join(","), /target must be inside repo/);
});

test("guard fails when target does not match repo/node_modules exactly", () => {
  const guard = evaluateCleanupGuards(row({ nodeModulesPath: "/tmp/demo-repo/sub/node_modules" }));
  assert.equal(guard.ok, false);
  assert.match(guard.reasons.join(","), /target must match repo\/node_modules/);
});

test("dry-run executes without touching fs", async () => {
  const result = await executeCleanup(row(), true);
  assert.equal(result.ok, true);
  assert.equal(result.dryRun, true);
  assert.equal(result.deletedPath, "/tmp/demo-repo/node_modules");
});

test("confirm token is explicit and repo-scoped", () => {
  assert.equal(buildConfirmToken(row()), "DELETE_NODE_MODULES demo-repo");
});

test("safety blocks unknown package manager", async () => {
  const assessment = await assessCleanupSafety(row({ manager: "unknown", hasLockfile: false }));
  assert.equal(assessment.ok, false);
  assert.equal(assessment.riskLevel, "high");
  assert.match(assessment.reasons.join(","), /unknown package manager/);
});

test("safety blocks yarn zero-install cache", async () => {
  const tempRoot = await fs.mkdtemp(path.join(os.tmpdir(), "cleanup-risk-"));
  const repoPath = path.join(tempRoot, "repo");
  await fs.mkdir(path.join(repoPath, ".yarn", "cache"), { recursive: true });
  const assessment = await assessCleanupSafety(
    row({
      repoPath,
      nodeModulesPath: path.join(repoPath, "node_modules"),
      manager: "yarn",
      hasLockfile: true
    })
  );
  assert.equal(assessment.ok, false);
  assert.match(assessment.reasons.join(","), /\.yarn\/cache/);
  await fs.rm(tempRoot, { recursive: true, force: true });
});

test("safety warns for recently active repo", async () => {
  const assessment = await assessCleanupSafety(row({ inactiveDays: 2 }));
  assert.equal(assessment.ok, true);
  assert.equal(assessment.riskLevel, "medium");
  assert.match(assessment.warnings.join(","), /recently active/);
});

File: internal/cleanup/cleanup_test.go
package cleanup

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/ericdahl-dev/repo-cleanup-tui/internal/scanner"
)

func row(overrides func(*scanner.Candidate)) scanner.Candidate {
	repoPath := "/tmp/demo-repo"
	c := scanner.Candidate{
		RepoPath:         repoPath,
		NodeModulesPath:  filepath.Join(repoPath, "node_modules"),
		Manager:          scanner.ManagerYarn,
		HasLockfile:      true,
		InactiveDays:     intPtr(120),
		Bytes:            1234,
		ReinstallCommand: "yarn install --immutable",
		Git:              scanner.GitContext{Branch: "main", Dirty: false},
	}
	if overrides != nil {
		overrides(&c)
	}
	return c
}

func intPtr(v int) *int { return &v }

func TestGuardPassesForRepoNodeModulesWithLockfile(t *testing.T) {
	guard := EvaluateGuards(row(nil))
	if !guard.OK {
		t.Fatalf("expected ok guard, reasons=%v", guard.Reasons)
	}
}

func TestGuardFailsWhenLockfileMissing(t *testing.T) {
	guard := EvaluateGuards(row(func(c *scanner.Candidate) { c.HasLockfile = false }))
	if guard.OK {
		t.Fatal("expected guard failure")
	}
	if !strings.Contains(strings.Join(guard.Reasons, ","), "missing lockfile") {
		t.Fatalf("reasons=%v", guard.Reasons)
	}
}

func TestGuardFailsWhenTargetNotNodeModules(t *testing.T) {
	guard := EvaluateGuards(row(func(c *scanner.Candidate) {
		c.NodeModulesPath = "/tmp/demo-repo/dist"
	}))
	if guard.OK || !strings.Contains(strings.Join(guard.Reasons, ","), "target must be node_modules") {
		t.Fatalf("reasons=%v", guard.Reasons)
	}
}

func TestGuardFailsWhenTargetOutsideRepo(t *testing.T) {
	guard := EvaluateGuards(row(func(c *scanner.Candidate) {
		c.NodeModulesPath = "/tmp/other/node_modules"
	}))
	if guard.OK || !strings.Contains(strings.Join(guard.Reasons, ","), "target must be inside repo") {
		t.Fatalf("reasons=%v", guard.Reasons)
	}
}

func TestGuardFailsWhenTargetNotExactRepoNodeModules(t *testing.T) {
	guard := EvaluateGuards(row(func(c *scanner.Candidate) {
		c.NodeModulesPath = "/tmp/demo-repo/sub/node_modules"
	}))
	if guard.OK || !strings.Contains(strings.Join(guard.Reasons, ","), "target must match repo/node_modules") {
		t.Fatalf("reasons=%v", guard.Reasons)
	}
}

func TestDryRunExecutesWithoutTouchingFS(t *testing.T) {
	result, err := Execute(row(nil), true)
	if err != nil {
		t.Fatal(err)
	}
	if !result.OK || !result.DryRun {
		t.Fatalf("result=%+v", result)
	}
	if result.DeletedPath != "/tmp/demo-repo/node_modules" {
		t.Fatalf("deletedPath=%q", result.DeletedPath)
	}
}

func TestConfirmTokenIsExplicitAndRepoScoped(t *testing.T) {
	got := BuildConfirmToken(row(nil))
	want := "DELETE_NODE_MODULES demo-repo"
	if got != want {
		t.Fatalf("got %q want %q", got, want)
	}
}

func TestSafetyBlocksUnknownPackageManager(t *testing.T) {
	assessment, err := AssessSafety(row(func(c *scanner.Candidate) {
		c.Manager = scanner.ManagerUnknown
		c.HasLockfile = false
	}))
	if err != nil {
		t.Fatal(err)
	}
	if assessment.OK || assessment.RiskLevel != RiskHigh {
		t.Fatalf("assessment=%+v", assessment)
	}
	if !strings.Contains(strings.Join(assessment.Reasons, ","), "unknown package manager") {
		t.Fatalf("reasons=%v", assessment.Reasons)
	}
}

func TestSafetyBlocksYarnZeroInstallCache(t *testing.T) {
	tempRoot := t.TempDir()
	repoPath := filepath.Join(tempRoot, "repo")
	if err := os.MkdirAll(filepath.Join(repoPath, ".yarn", "cache"), 0o755); err != nil {
		t.Fatal(err)
	}
	assessment, err := AssessSafety(row(func(c *scanner.Candidate) {
		c.RepoPath = repoPath
		c.NodeModulesPath = filepath.Join(repoPath, "node_modules")
		c.Manager = scanner.ManagerYarn
		c.HasLockfile = true
	}))
	if err != nil {
		t.Fatal(err)
	}
	if assessment.OK || !strings.Contains(strings.Join(assessment.Reasons, ","), ".yarn/cache") {
		t.Fatalf("assessment=%+v", assessment)
	}
}

func TestSafetyWarnsForRecentlyActiveRepo(t *testing.T) {
	assessment, err := AssessSafety(row(func(c *scanner.Candidate) {
		c.InactiveDays = intPtr(2)
	}))
	if err != nil {
		t.Fatal(err)
	}
	if !assessment.OK || assessment.RiskLevel != RiskMedium {
		t.Fatalf("assessment=%+v", assessment)
	}
	if !strings.Contains(strings.Join(assessment.Warnings, ","), "recently active") {
		t.Fatalf("warnings=%v", assessment.Warnings)
	}
}
